go/tools/builders/nogo_main.go

```diff
@@ -202,6 +202,55 @@ func checkPackage(analyzers []*analysis.Analyzer, packagePath string, packageFil
 		act.pkg = pkg
 	}
 
+	// Process nolint directives similar to golangci-lint.
+	for _, f := range pkg.syntax {
+		// First, gather all comments and apply a range to the entire comment block.
+		// This will cover inline comments and allow us to detect comments above a
+		// node later.
+		for _, group := range f.Comments {
+			for _, comm := range group.List {
+				linters, ok := parseNolint(comm.Text)
+				if !ok {
+					continue
+				}
+				nl := &Range{
+					from: pkg.fset.Position(group.Pos()),
+					to:   pkg.fset.Position(group.End()).Line,
+				}
+				for analyzer, act := range actions {
+					if linters.Contains(analyzer.Name) {
+						act.nolint = append(act.nolint, nl)
+					}
+				}
+			}
+		}
+
+		// For each node in the ast, check if the previous line is covered by a
+		// range for the linter and expand it.
+		ast.Inspect(f, func(n ast.Node) bool {
```

Comment:
I am not sure why we are checking the previous line of an AST node and not the first line. I think future maintainers might be confused as well.

It's best if you could provide a separate test(s) that would help demonstrate the range of what would be covered and NOT covered when a `//nolint` comment is made.

Reply:
I'm doing a lot of same logic as golangci-lint but did not provide a lot of context here.

The way I reasoned about this is there are 2 cases we are trying to handle:
1. Inline comments - these apply to the same line as the ast node
2. Comments above the block they apply to - this is the case handled here

The idea is that something like
```
//nolint
complex(stmt.Func())
```
Will apply the `nolint` directive to everything within the following line

A more complex example:
```
//nolint
return Struct{
  foo: bar,
}, nil
```

The complex example should cover any issues that arise in the _entire_ return block. I can add more tests to show these examples.